source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -61,10 +114,954 @@ QString AMPIC887Controller::status() const
+	return QList<int>();
+}
+
+QString AMPIC887Controller::recorderOptionsString() const
+{
+	return controllerState_->dataRecorderState()->availableParameters();
+}
+
+AMGCS2::DataRecordTrigger AMPIC887Controller::recordTrigger() const
+{
+	return controllerState_->dataRecorderState()->recordTrigger();
+}
+
+bool AMPIC887Controller::setRecordTrigger(AMGCS2::DataRecordTrigger recordTrigger)
+{
+	if(controllerState_->dataRecorderState()->recordTrigger() == recordTrigger) {
+		return true;
+	}
+
+	AMGCS2SetRecordTriggerSourceCommand setRecordTriggerCommand(recordTrigger);
+	runCommand(&setRecordTriggerCommand);
+
+	if(setRecordTriggerCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setRecordTriggerCommand.lastError());
+		return false;
+	}
+
+	controllerState_->dataRecorderState()->setRecordTrigger(recordTrigger);
+
+	return true;
+}
+
+bool AMPIC887Controller::isAxisReferenced(AMGCS2::Axis axis) const
+{
+	return controllerState_->hexapodState()->referencedState(axis);
+}
+
+AMPIC887AxisMap<bool> AMPIC887Controller::axisReferencedStates(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->referencedStates().values(axes);
+}
+
+bool AMPIC887Controller::areAxesReferenced() const
+{
+	return controllerState_->hexapodState()->areAllAxesReferenced();
+}
+
+bool AMPIC887Controller::isInServoMode() const
+{
+	return controllerState_->hexapodState()->isInServoMode();
+}
+
+bool AMPIC887Controller::setServoMode(bool servoMode)
+{
+	if(controllerState_->hexapodState()->isInServoMode() == servoMode) {
+		return true;
+	}
+
+	AMGCS2SetServoModeCommand setServoModeCommand(servoMode);
+	runCommand(&setServoModeCommand);
+
+	if(setServoModeCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setServoModeCommand.lastError());
+		return false;
+	}
+
+	controllerState_->hexapodState()->setServoMode(servoMode);
+	return true;
+}
+
+double AMPIC887Controller::stepSize(AMGCS2::Axis axis)
+{
+	return controllerState_->hexapodState()->stepSize(axis);
+}
+
+bool AMPIC887Controller::setStepSize(AMGCS2::Axis axis, double stepSize)
+{
+	AMPIC887AxisMap<double> singleValue;
+	singleValue.insert(axis, stepSize);
+	return setStepSizes(singleValue);
+}
+
+AMPIC887AxisMap<double> AMPIC887Controller::stepSizes(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->stepSizes().values(axes);
+}
+
+bool AMPIC887Controller::setStepSizes(const AMPIC887AxisMap<double> &stepSizes)
+{
+	if(controllerState_->hexapodState()->stepSizes() == stepSizes) {
+		return true;
+	}
+
+	AMGCS2SetStepSizeCommand setStepSizeCommand(stepSizes);
+	runCommand(&setStepSizeCommand);
+
+	if(setStepSizeCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setStepSizeCommand.lastError());
+		return false;
+	}
+
+	controllerState_->hexapodState()->setStepSizes(stepSizes);
+	return true;
+}
+
+double AMPIC887Controller::systemVelocity() const
+{
+	return controllerState_->hexapodState()->velocity();
```

Comment:
The controllerState_ is a value member, so that can't ever be null, and while the hexapodState() is a reference, it's a reference whose lifetime is tied to that of the controllerState_.